### PR TITLE
Result Filtering

### PR DESCRIPTION
**Pull request by fneu**

### CODE REVIEW

**Comment by sils1297** [comment]
@fneu progress? please also fix the gitmate review :)

File: coalib/results/SourceRange.py [hunk 1]
```diff
@@ -59,4 +59,7 @@ def file(self):
         return self.start.file
 
     def overlaps(self, other):
+        if self.file != other.file:
+            return False
+
         return self.start <= other.end and self.end >= other.start
```
Comment (sils1297) [inline coalib/results/SourceRange.py]: hey, just make this `self.file != other.file and ...`

File: coalib/results/ResultFilter.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+
+    :param original_result: A result of the old files
+    :param modified_result: A result of the new files
+    :return:                Boolean value whether or not the properties match
+    """
+    # origin might be a class or class name
+    original_origin = isinstance(original_result.origin, str) and \
+        original_result.origin or original_result.origin.__name__()
+    modified_origin = isinstance(modified_result.origin, str) and \
+        modified_result.origin or modified_result.origin.__name__()
```
Comment (sils1297) [inline coalib/results/ResultFilter.py]: the origin is always a string, you can create this with a class for convenience but internally it's always a string

File: coalib/results/ResultFilter.py [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+        return False
+
+    elif original_result.severity != modified_result.severity:
+        return False
+
+    elif original_result.debug_msg != modified_result.debug_msg:
+        return False
+
+    else:
+        return True
```
Comment (sils1297) [inline coalib/results/ResultFilter.py]: why not simply `return original_result.origin == modified_result.origin and original_result..message == ...`?

File: coalib/results/ResultFilter.py [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+    :param original_result_diff_dict: diff for each file for this result
+    :param modified_result:    A result of the new files
+    :param modified_result_diff_dict: guess
+    :return:                     Boolean value whether the SourceRanges match
+    """
+    for file_name in original_file_dict:
+
+        try:  # fails if the affected range of the result get's modified
+            original_total_diff = \
+                diff_dict[file_name] + original_result_diff_dict[file_name]
```
Comment (sils1297) [inline coalib/results/ResultFilter.py]: try to use brackets instead of backslashes, otherwise mischa will kill you.

Comment (sils1297) [inline coalib/results/ResultFilter.py]: at least yell badly

File: coalib/results/ResultFilter.py [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+    Checks whether the Diffs of two results describe the same changes
+
+    :param original_file_dict: Dict of lists of file contents before changes
+    :param modified_file_dict: Dict of lists of file contents after changes
+    :param diff_dict:          Dict of diffs describing the change in each file
+    :param original_result:    A result of the old files
+    :param modified_result:    A result of the new files
+    :return:                   Boolean value whether the Diffs match
+    """
+    return False # pragma: no cover
```
Comment (sils1297) [inline coalib/results/ResultFilter.py]: huh? What's this :D

File: coalib/results/ResultFilter.py [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+            uniques.append(m_r)
+
+        elif True not in [source_ranges_match(
+                original_file_dict,
+                modified_file_dict,
+                diffs_dict,
+                o_r,
+                orig_result_diff_dict_dict[o_r],
+                m_r,
+                mod_result_diff_dict_dict[m_r]) for o_r in original_results]:
```
Comment (sils1297) [inline coalib/results/ResultFilter.py]: you should only consider the ones where the basics matched before here!

1. For performance
2. They both must match, there might be another result which has other basics but the same diff.

Comment (sils1297) [inline coalib/results/ResultFilter.py]: so you probably want one iteration over all modified, inside an iteration over all originals and then do the checks sequencially.

File: coalib/results/ResultFilter.py (hunk 3 above)
Comment (fneu) [inline coalib/results/ResultFilter.py]: One Line to rule them all,
one line to find them,
one line to bring then all and
in the darkness bind them

Tbh. I just found this more expressive although one can make a case against 
using `!=` all that much

File: coalib/results/ResultFilter.py (hunk 5 above)
Comment (fneu) [inline coalib/results/ResultFilter.py]: Infrastructure... I can take it out now but I will need it anyways.

File: coalib/results/ResultFilter.py (hunk 6 above)
Comment (fneu) [inline coalib/results/ResultFilter.py]: Yeah, I thought I could add the second check like this... But that is 
complete rubbish isn't it? I'm amazed this works for all my tests...

File: coalib/results/ResultFilter.py (hunk 3 above)
Comment (sils1297) [inline coalib/results/ResultFilter.py]: maybe even a loop: `all(getattr(original_result, member) == getattr(modified_result, member) for member in ['origin', 'message', ...])`?

File: coalib/results/ResultFilter.py (hunk 6 above)
Comment (sils1297) [inline coalib/results/ResultFilter.py]: just fix it in a performance efficient way, this has a very bad complexity `O(n*m)`, if basics are not fit there's no need to match the rest which takes a lot of time

File: coalib/results/ResultFilter.py (hunk 5 above)
Comment (sils1297) [inline coalib/results/ResultFilter.py]: especially the pragma no cover is funny :P

File: coalib/results/ResultFilter.py (hunk 3 above)
Comment (fneu) [inline coalib/results/ResultFilter.py]: now that is pretty elegant, I'm impressed!

File: coalib/results/ResultFilter.py (hunk 5 above)
Comment (fneu) [inline coalib/results/ResultFilter.py]: well, otherwise I would have needed to call it in filter_results... :p

Comment (sils1297) [inline coalib/results/ResultFilter.py]: ah interesting, I would have guessed that you need to do the pragma for the whole function... well then take it out until you need it ok?

**Comment by sils1297** [comment]
@fneu I can close this, right?

**Comment by sils1297** [comment]
restore and reopen if needed

**Pull request closed without merging**